Adding SelfJoin anti pattern

paraschavda2411 opened the pull request.

franklinWhaite commented:
1. Please resolve the conflicts in main
2. Please add the an example of a self join the the [sample folder](https://github.com/GoogleCloudPlatform/bigquery-antipattern-recognition/tree/main/samples/queries/input) and the [sample csv](https://github.com/GoogleCloudPlatform/bigquery-antipattern-recognition/blob/main/samples/csv/input/input_queries.csv)

franklinWhaite commented:
Please add the following test case:


```
SELECT 
  date(t1.trip_start_timestamp) dt,
  COUNT(DISTINCT t1.unique_key) ct,
  SUM(t2.ct) ct_month
FROM 
  `bigquery-public-data.chicago_taxi_trips.taxi_trips` t1
LEFT JOIN
  (SELECT 
    FORMAT_DATE("%Y-%m", trip_start_timestamp) month, 
    COUNT(DISTINCT unique_key) ct
  FROM `bigquery-public-data.chicago_taxi_trips.taxi_trips`
  GROUP BY month
  ) t2 
    ON month = FORMAT_DATE("%Y-%m", t1.trip_start_timestamp)
WHERE
  FORMAT_DATE('%Y-%m', t1.trip_start_timestamp) = '2023-02'
  AND t2.month = '2023-02'
GROUP BY
  dt
ORDER BY
  dt desc
```

franklinWhaite commented on bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifySelfJoinVisitor.java:
```diff
@@ -0,0 +1,78 @@
+
+    // Loop through all the identifiers in the table path expression.
+    public void getIdentifier(ASTNodes.ASTTablePathExpression tablePathExpression, String joinSide) {
+//        System.out.println("Table Path Expression ==> "+ tablePathExpression);
```
```suggestion

```

paraschavda2411 replied:
Done

franklinWhaite commented:
**Bug**
The following query is incorrectly labeled as a case of the self-join anti-pattern

`
```
SELECT
    t1.col1
FROM 
    table1 t1 
JOIN 
    table2 t2 on t1.col1 = t2.col2
UNION ALL
SELECT
    t2.col1
FROM 
    table2 t2 
JOIN 
    table3 t3 on t1.col1 = t2.col2
```

paraschavda2411 commented:
Resolved the bug
Added the test case suggested
Resolved the conflicts in main
Added sample csv, queries and outputs
Updated the read-me content 

Please review and let me know if anything

franklinWhaite commented on README.md:
```diff
@@ -469,5 +469,38 @@ WhereOrder: LIKE filter in line 8 precedes a more selective filter.
 
+
+## Anti Pattern 10: Using an analytic functions over self-joins
+Example:
```
Can we please use a simpler example here

```
select 
  t1.*, 
  t2.value as prev_value
from 
  table t1
left join 
  table t2
  on 
    t1.id = t2.id
    and t1.date - 1 = t2.date
```

paraschavda2411 replied:
Gave simple examples

franklinWhaite commented on bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifySelfJoinVisitor.java:
```diff
@@ -0,0 +1,86 @@
+                            table = identifier.getIdString().toLowerCase();
+                            tableIdMap.put(joinSide,table);
+                            /*System.out.println("Table Map - > " + tableIdMap);
+                            System.out.println("==========================================");
```
Please remove debugging comments before merging

paraschavda2411 replied:
Removed debugging comments

franklinWhaite commented on samples/csv/input/input_queries.csv:
```diff
@@ -91,3 +91,26 @@ where
      and repo_name = 'cdnjs/cdnjs';
 "
+
+9, "SELECT
```
Can we please use a simpler example here. (See comment in readme)

paraschavda2411 replied:
Done

franklinWhaite commented on samples/queries/input/SelfJoin.sql:
```diff
@@ -0,0 +1,21 @@
+SELECT
```
Can we please use a simpler example here. (See readme comment)

paraschavda2411 replied:
used simple examples
